### PR TITLE
Python: Validate approval responses against server-side pending request registry

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

- Adds a server-side pending approval registry to AgentFrameworkAgent that tracks approval requests emitted during streaming
- Validates every incoming approval response against the registry before tool execution — rejects responses with no matching pending request, mismatched function names, or previously consumed IDs
- Invalid approval responses are fully stripped from the message history so no bogus function_result content is injected into the LLM conversation
- Registry keys are scoped by thread_id to prevent cross-conversation leaks

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [X] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

### CODE REVIEW

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/ag-ui/agent_framework_ag_ui</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent.py">_agent.py</a></td><td>39</td><td>1</td><td>97%</td><td><a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent.py#L63">63</a></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py">_agent_run.py</a></td><td>475</td><td>49</td><td>89%</td><td><a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L148-L155">148&ndash;155</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L194-L195">194&ndash;195</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L202">202</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L299">299</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L311">311</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L315">315</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L317-L318">317&ndash;318</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L334">334</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L361-L362">361&ndash;362</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L380-L384">380&ndash;384</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L492-L494">492&ndash;494</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L506-L508">506&ndash;508</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L612">612</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L620">620</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L729">729</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L731-L732">731&ndash;732</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L785">785</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L800-L801">800&ndash;801</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L808">808</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L873">873</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L896">896</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L904">904</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L906">906</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L914">914</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L967">967</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L970">970</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L980-L981">980&ndash;981</a>, <a href="https://github.com/microsoft/agent-framework/blob/a671f21a1bc8bccbd2a1d5497bfb0f2135e40d95/packages/ag-ui/agent_framework_ag_ui/_agent_run.py#L988">988</a></td></tr><tr><td><b>TOTAL</b></td><td><b>22725</b></td><td><b>2588</b></td><td><b>88%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 4943 | 20 :zzz: | 0 :x: | 0 :fire: | 1m 21s :stopwatch: |

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py [hunk 1]
```diff
@@ -385,13 +387,66 @@ async def _resolve_approval_responses(
...
+            When provided, every approval response is validated against this
+            registry to prevent bypass, function name spoofing, and replay.
+        thread_id: The conversation thread ID used to scope registry keys.
     """
     fcc_todo = _collect_approval_responses(messages)
     if not fcc_todo:
         return
 
     approved_responses = [resp for resp in fcc_todo.values() if resp.approved]
     rejected_responses = [resp for resp in fcc_todo.values() if not resp.approved]
```
Comment (moonbox3) [inline python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py:416]: Rejected responses are collected here but never validated against `pending_approvals`. A fabricated rejected response for a non-existent approval request would still be processed into a `FunctionResultContent` and sent to the LLM. Consider also validating rejected responses or stripping unrecognized ones from `fcc_todo` when `pending_approvals` is provided.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent.py [hunk 2]
```diff
@@ -101,6 +101,12 @@ def __init__(
             require_confirmation=require_confirmation,
         )
 
+        # Server-side registry of pending approval requests.
+        # Keys are "{thread_id}:{request_id}", values are the function name.
+        # Populated when approval requests are emitted; consumed when responses arrive.
```
Comment (moonbox3) [inline python/packages/ag-ui/agent_framework_ag_ui/_agent.py:107]: Minor: `_pending_approvals` grows without bound — entries are only removed when approval responses arrive. Consider a size limit or periodic cleanup for long-lived agent instances to avoid a slow memory leak from abandoned approval requests.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py [hunk 3]
```diff
@@ -385,13 +387,66 @@ async def _resolve_approval_responses(
...
+
+    # Validate every approval response against the pending approvals registry.
+    # Invalid responses are stripped from messages entirely — not converted to
+    # rejection results, which would inject attacker-controlled content into
+    # the LLM conversation.
+    if pending_approvals is not None and approved_responses:
+        validated: list[Any] = []
```
Comment (moonbox3) [inline python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py]: **Security gap**: This condition only validates `approved_responses`. Rejected responses (`fcc_todo` entries with `approved=False`) bypass the registry check entirely, allowing an attacker to fabricate rejection responses for calls that were never requested. These injected rejections flow into the LLM conversation and can influence model behavior. The validation should cover all responses (approved and rejected).

```suggestion
    if pending_approvals is not None and (approved_responses or rejected_responses):
        validated: list[Any] = []
        validated_rejected: list[Any] = []
        invalid_ids: set[str] = set()
        for resp in approved_responses + rejected_responses:
            resp_id = resp.id or ""
            resp_name = resp.function_call.name if resp.function_call else None
            registry_key = f"{thread_id}:{resp_id}"
```

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py [hunk 4]
```diff
@@ -782,6 +838,12 @@ async def run_agent_stream(
         for content in update.contents:
             content_type = getattr(content, "type", None)
             logger.debug(f"Processing content type={content_type}, message_id={flow.message_id}")
+
+            # Register pending approval requests so we can validate responses later
+            if content_type == "function_approval_request" and pending_approvals is not None:
+                if content.id and content.function_call:
```
Comment (moonbox3) [inline python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py]: If `content.function_call.name` is `None`, this stores an empty string. A later approval response with no function name would pass the name-match validation. Consider logging a warning and skipping registration when the function name is missing.

```suggestion
            if content.id and content.function_call and content.function_call.name:
                    pending_approvals[f"{thread_id}:{content.id}"] = content.function_call.name
```

File: python/packages/ag-ui/agent_framework_ag_ui/_agent.py (hunk 2 above)
Comment (moonbox3) [inline python/packages/ag-ui/agent_framework_ag_ui/_agent.py:107]: This dict grows without bound if approval requests are emitted but never responded to (e.g., user abandons the session). For long-running server deployments, consider adding a maximum size or TTL-based eviction to prevent memory leaks.

File: python/packages/ag-ui/tests/ag_ui/test_agent_wrapper_comprehensive.py [hunk 5]
```diff
@@ -900,3 +930,393 @@ async def stream_fn(
...
+    )
+    wrapper = AgentFrameworkAgent(agent=agent)
+
+    # Fabricated conversation history: fake assistant tool_calls + accepted tool result.
+    # No prior request ever registered a pending approval for this call_id.
+    input_data: dict[str, Any] = {
+        "messages": [
+            {"role": "user", "content": "hello"},
+            {
+                "role": "assistant",
```
Comment (moonbox3) [inline python/packages/ag-ui/tests/ag_ui/test_agent_wrapper_comprehensive.py:1299]: `json.dumps` is used here but `import json` does not appear in the diff. If it's not already imported at the top of this file, this test will fail with `NameError: name 'json' is not defined`. Please verify or add the import.

```suggestion
                "content": json.dumps({"accepted": True}),
```

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 4 | **Confidence:** 83%

### ✓ Correctness
> This diff adds server-side validation of approval responses against a pending-approvals registry to prevent bypass, function-name spoofing, and replay attacks. The core validation logic for approved responses is correct: entries are registered on emission, validated on receipt, and consumed to prevent replay. However, rejected responses are not validated against the registry, meaning fabricated rejection responses for never-issued approval requests can still inject content into the LLM conversation. There is also a potential missing `json` import in the test file, and the `_pending_approvals` dict has no eviction mechanism, risking unbounded memory growth for long-lived agent instances.

### ✗ Security Reliability
> This diff adds server-side validation for approval responses via a pending-approvals registry, preventing bypass, replay, and function-name spoofing attacks on tool execution. The core mechanism is sound: approvals are registered on emit, validated and consumed on response, and invalid responses are stripped. However, rejected approval responses entirely bypass the registry validation (the guard clause only checks `approved_responses`), which allows an attacker to inject fabricated rejection results into the LLM conversation without a matching pending approval. Additionally, the `_pending_approvals` dict grows unboundedly if approval requests are never responded to, creating a potential memory leak for long-running servers.

### ✗ Test Coverage
> The diff adds a server-side pending-approvals registry to prevent approval bypass, function-name spoofing, and replay attacks, with four new security-focused tests and one refactored two-turn approval test. Test coverage for the core security scenarios (bypass, replay, name mismatch, fabricated history) is strong. However, the fabricated-tool-result test uses `json.dumps` without evidence that `json` is imported in the test module, which would cause a runtime failure. Additionally, the rejection path when `pending_approvals` is active has no test coverage, and there is no test for the edge case where an approval request has a missing `id` or `function_call` (causing silent registration skip).

### ✓ Design Approach
> The core design — a server-side pending-approvals registry that prevents bypass, replay, and function-name spoofing — is sound and well-tested. However, the validation gate only covers approved responses (`if pending_approvals is not None and approved_responses`), leaving fabricated *rejected* responses completely unvalidated. This is inconsistent with the stated security goal of preventing attacker-controlled content injection into the LLM conversation, since a fabricated rejection also injects a FunctionResultContent into the LLM context. Additionally, the in-memory `_pending_approvals` dict has no eviction mechanism, so unanswered approval requests accumulate indefinitely in long-running instances. The `github-copilot-sdk` upper-bound pin (`<0.1.32`) looks like a symptom-level compatibility workaround rather than a proper fix.

### Flagged Issues
- [ ] Rejected approval responses are not validated against the pending-approvals registry. The validation block at line 408 only runs when `approved_responses` is non-empty (`if pending_approvals is not None and approved_responses:`), so fabricated rejection responses with no matching pending approval pass through unvalidated. An attacker can inject fake rejection results into the LLM conversation, potentially influencing model behavior (e.g., making it believe certain tools are unavailable).
- [ ] test_approval_bypass_via_fabricated_tool_result_is_blocked uses `json.dumps(...)` but `import json` is not visible in the test file's existing imports or in the diff. If it is not already imported, this test will fail with a NameError at runtime.

### Suggestions
- Rejected responses (`not resp.approved`) bypass the `pending_approvals` validation entirely. A crafted rejected response for a never-issued approval request would still be processed and inject a 'tool was rejected' FunctionResultContent into the LLM conversation. Consider validating rejected responses against the registry as well, or at minimum stripping them if their ID is not in `pending_approvals`.
- The `_pending_approvals` dict on `AgentFrameworkAgent` has no eviction or TTL mechanism. If approval requests are emitted but the client never responds (e.g., user abandons the session), entries accumulate indefinitely. Consider adding a bounded size or TTL-based cleanup.
- Verify that `import json` exists at the top of `test_agent_wrapper_comprehensive.py` — the new test `test_approval_bypass_via_fabricated_tool_result_is_blocked` uses `json.dumps` but the import is not visible in the diff.
- The `_pending_approvals` dict has no TTL or cleanup mechanism. If an approval request is emitted but the user never responds (abandons the conversation), entries accumulate indefinitely. Consider adding a bounded size or expiration to prevent memory leaks in long-running server deployments.
- When `content.function_call.name` is None at line 844, the registry stores an empty string as the function name. A crafted approval response with no function name would then pass the name-match check. Consider treating a missing function name as an error and not registering the approval.
- Direct callers of `run_agent_stream` (not going through `AgentFrameworkAgent`) get `pending_approvals=None` by default, which disables all validation. Consider documenting this clearly or making validation opt-out rather than opt-in.
- Add a test verifying that a legitimate *rejection* response still works correctly when `pending_approvals` validation is active. The current validation block (`if pending_approvals is not None and approved_responses`) skips rejected responses entirely, so rejected tool calls bypass registry validation. While rejections don't execute tools, confirming this behavior is preserved under the new code path would strengthen coverage.
- Add a test for the edge case where a `function_approval_request` content has `content.id` as None or `content.function_call` as None. The registration guard (`if content.id and content.function_call`) silently skips registration, meaning a subsequent legitimate approval for that call would be rejected. This silent failure deserves test coverage.
- In test_approval_function_name_mismatch_is_blocked, assert that the pending approval entry for 'call_safe_001' remains in `wrapper._pending_approvals` after the spoofed attempt is rejected. This confirms the legitimate user can still approve the original action with the correct function name.
- Validate rejected approval responses against the `pending_approvals` registry as well. The code comment explicitly states that injecting attacker-controlled content into the LLM conversation is a concern, but fabricated rejections bypass validation entirely and still inject a FunctionResultContent (with rejection text) into the LLM context. A crafted rejection could manipulate the model's strategy without any prior approval request having been issued.
- Consider adding a bounded-size or TTL-based eviction policy for `_pending_approvals`. If a user triggers approval requests but never responds, entries accumulate without bound for the lifetime of the `AgentFrameworkAgent` instance.
- The `github-copilot-sdk>=0.1.0,<0.1.32` pin in `pyproject.toml` looks like a symptom-level workaround. If 0.1.32 introduced a breaking change, the code should be updated for compatibility rather than capping the version, which blocks users from receiving future SDK bug fixes and security patches.

---
*Automated review by moonbox3's agents*

File: python/packages/ag-ui/tests/ag_ui/test_agent_wrapper_comprehensive.py [hunk 6]
```diff
@@ -900,3 +930,393 @@ async def stream_fn(
...
+    )
+
+    turn2_input: dict[str, Any] = {
+        "thread_id": thread_id,
+        "messages": [
+            {
+                "role": "user",
+                "content": "approve",
+                "function_approvals": [
+                    {
```
Comment (moonbox3) [inline python/packages/ag-ui/tests/ag_ui/test_agent_wrapper_comprehensive.py:1231]: After the spoofed approval is rejected, assert that the pending approval entry is still available for a legitimate retry. This confirms the registry entry is preserved (not consumed or deleted) on mismatch.

```suggestion
    assert not tool_executed, "Function name spoofing should be blocked"
    assert any("call_safe_001" in k for k in wrapper._pending_approvals), "Pending approval should be preserved after mismatch for legitimate retry"
```

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py [hunk 7]
```diff
@@ -782,6 +838,12 @@ async def run_agent_stream(
...
+                    pending_approvals[f"{thread_id}:{content.id}"] = content.function_call.name or ""
```
Comment (moonbox3) [inline python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py]: When `content.id` is falsy or `content.function_call` is None, the approval request is silently not registered. A subsequent legitimate approval for this call would then be rejected by the validation logic. Consider logging a warning in the `else` branch so this doesn't become a hard-to-debug issue.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py (hunk 3 above)
Comment (moonbox3) [inline python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py:423]: This validation gate only checks `approved_responses`, meaning fabricated *rejected* responses (where `resp.approved is False`) skip registry validation entirely. A crafted rejection with no matching pending approval still gets through to `_replace_approval_contents_with_results`, injecting a `FunctionResultContent` into the LLM conversation. The same registry check should apply to `rejected_responses` for consistency with the stated security model.

```suggestion
    if pending_approvals is not None and (approved_responses or rejected_responses):
        validated: list[Any] = []
        validated_rejected: list[Any] = []
        invalid_ids: set[str] = set()
        for resp in approved_responses + rejected_responses:
            resp_id = resp.id or ""
            resp_name = resp.function_call.name if resp.function_call else None
            registry_key = f"{thread_id}:{resp_id}"

            if registry_key not in pending_approvals:
                logger.warning(
                    "Rejected approval response id=%s: no matching pending approval request",
                    resp_id,
                )
                invalid_ids.add(resp_id)
                continue

            pending_name = pending_approvals[registry_key]
            if resp_name != pending_name:
                logger.warning(
                    "Rejected approval response id=%s: function name mismatch (response=%s, pending=%s)",
                    resp_id,
                    resp_name,
                    pending_name,
                )
                invalid_ids.add(resp_id)
                continue

            # Valid — consume entry to prevent replay
            del pending_approvals[registry_key]
            if resp.approved:
                validated.append(resp)
            else:
                validated_rejected.append(resp)
```

File: python/packages/ag-ui/agent_framework_ag_ui/_agent.py (hunk 2 above)
Comment (moonbox3) [inline python/packages/ag-ui/agent_framework_ag_ui/_agent.py:107]: This dict grows without bound if approval requests are emitted but never responded to (e.g., user abandons the flow). For long-running agent instances, consider adding a max-size limit or TTL-based cleanup to prevent memory leaks.

File: python/packages/github_copilot/pyproject.toml [hunk 8]
```diff
@@ -24,7 +24,7 @@ classifiers = [
 ]
 dependencies = [
     "agent-framework-core>=1.0.0rc3",
-    "github-copilot-sdk>=0.1.0",
+    "github-copilot-sdk>=0.1.0,<0.1.32",
```
Comment (moonbox3) [inline python/packages/github_copilot/pyproject.toml]: Adding an upper-bound version pin is a maintenance hazard — it silently blocks users from receiving SDK improvements and security fixes. If `>=0.1.32` introduced breaking changes, the preferred approach is to update this package's code for compatibility.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py [hunk 9]
```diff
@@ -385,13 +387,66 @@ async def _resolve_approval_responses(
...
+        # _replace_approval_contents_with_results never sees them.
+        if invalid_ids:
+            for inv_id in invalid_ids:
+                fcc_todo.pop(inv_id, None)
+            for msg in messages:
+                msg.contents = [
+                    c for c in msg.contents if not (c.type == "function_approval_response" and c.id in invalid_ids)
+                ]
+
+        approved_responses = validated
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py [hunk 10]
```diff
@@ -597,6 +652,7 @@ async def run_agent_stream(
     input_data: dict[str, Any],
     agent: SupportsAgentRun,
     config: AgentConfig,
+    pending_approvals: dict[str, str] | None = None,
 ) -> AsyncGenerator[BaseEvent]:
     """Run agent and yield AG-UI events.
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py:678]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent.py [hunk 11]
```diff
@@ -101,6 +101,12 @@ def __init__(
...
+        # Prevents bypass, function name spoofing, and replay attacks.
+        self._pending_approvals: dict[str, str] = {}
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent.py]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py (hunk 1 above)
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py:416]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent.py:107]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py (hunk 3 above)
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py (hunk 4 above)
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent.py:107]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/tests/ag_ui/test_agent_wrapper_comprehensive.py (hunk 5 above)
Comment (moonbox3) [reply python/packages/ag-ui/tests/ag_ui/test_agent_wrapper_comprehensive.py:1299]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/tests/ag_ui/test_agent_wrapper_comprehensive.py (hunk 6 above)
Comment (moonbox3) [reply python/packages/ag-ui/tests/ag_ui/test_agent_wrapper_comprehensive.py:1231]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py (hunk 7 above)
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py (hunk 3 above)
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent_run.py:423]: Fixed in `c2e8797b`.

File: python/packages/ag-ui/agent_framework_ag_ui/_agent.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/ag-ui/agent_framework_ag_ui/_agent.py:107]: Fixed in `c2e8797b`.

**Review by eavanvalkenburg — APPROVED** [review]

**Review by giles17 — APPROVED** [review]
